src/leo_nlm_proto4_server.erl

```diff
@@ -0,0 +1,314 @@
+
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2]).
+
+-export([nlmproc4_null_4/2,
+         nlmproc4_test_4/3,
+         nlmproc4_lock_4/3,
+         nlmproc4_cancel_4/3,
+         nlmproc4_unlock_4/3
+        ]).
+
+%% nlmproc4_granted_4/3,
+%% nlmproc4_test_msg_4/3,
+%% nlmproc4_lock_msg_4/3,
+%% nlmproc4_cancel_msg_4/3,
+%% nlmproc4_unlock_msg_4/3,
+%% nlmproc4_granted_msg_4/3,
+%% nlmproc4_test_res_4/3,
+%% nlmproc4_lock_res_4/3,
+%% nlmproc4_cancel_res_4/3,
+%% nlmproc4_unlock_res_4/3,
+%% nlmproc4_granted_res_4/3,
+%% nlmproc4_share_4/3,
+%% nlmproc4_unshare_4/3,
+%% nlmproc4_nm_lock_4/3,
+%% nlmproc4_free_all_4/3
+
+-record(lock_record,{
+          start :: non_neg_integer(),
+          till  :: integer(),
+          owner :: binary(),
+          excl  :: boolean()
+         }).
+-define(NLM_LOCK_ETS, nlm_lock_ets).
+
+-define(NLM4_GRANTED,               'NLM4_GRANTED').
+-define(NLM4_DENIED,                'NLM4_DENIED').
+-define(NLM4_DENIED_NOLOCKS,        'NLM4_DENIED_NOLOCKS').
+-define(NLM4_BLOCKED,               'NLM4_BLOCKED').
+-define(NLM4_DENIED_GRACE_PERIOD,   'NLM4_DENIED_GRACE_PERIOD').
+-define(NLM4_DEADLCK,               'NLM4_DEADLCK').
+-define(NLM4_ROFS,                  'NLM4_ROFS').
+-define(NLM4_STALE_FH,              'NLM4_STALE_FH').
+-define(NLM4_FBIG,                  'NLM4_FBIG').
+-define(NLM4_FAILED,                'NLM4_FAILED').
+
+%% ---------------------------------------------------------------------
+%% API
+%% ---------------------------------------------------------------------
+%% @doc Called only once from a parent rpc server process to initialize this module
+%%      during starting a leo_storage server.
+-spec(init(any()) -> {ok, any()}).
+init(_Args) ->
+    ets:new(?NLM_LOCK_ETS, [set, named_table, public]),
+    {ok, void}.
+
+handle_call(Req, _From, S) ->
+    ?debug("handle_call", "req:~p from:~p", [Req, _From]),
+    {reply, [], S}.
+
+handle_cast(Req, S) ->
+    ?debug("handle_cast", "req:~p", [Req]),
+    {reply, [], S}.
+
+handle_info(Req, S) ->
+    ?debug("handle_info", "req:~p", [Req]),
+    {noreply, S}.
+
+terminate(_Reason, _S) ->
+    ok.
+
+%% ---------------------------------------------------------------------
+%% API
+%% ---------------------------------------------------------------------
+%% @doc
+nlmproc4_null_4(_Clnt, State) ->
+    ?debug("NLM_NULL", "from:~p", [_Clnt]),
+    {reply, [], State}.
+
+%% @doc
+nlmproc4_test_4({Cookie, IsExclusive, Lock} = TestArgs, Clnt, State) ->
+    ?debug("NLM_TEST", "Args:~p, from:~p", [TestArgs, Clnt]),
+    {_CallerName, FileHandler, _Owner, _Svid, _Offset, _Length} = Lock,
+    NewLock = convert_lock(Lock, IsExclusive), 
+    LockRet = case test_lock(FileHandler, NewLock) of
+                  {ok, _} ->
+                      ?NLM4_GRANTED;
+                  _ ->
+                      ?NLM4_DENIED
+              end,
+    {reply, {Cookie, {LockRet}}, State}.
```

Comment:
It seems that `nlm4_holder` should be returned when `nlm4_stats` is `NLM4_DENIED`